pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -70,20 +76,42 @@ def get_profiled_train_dataloader(self, train_dataloader) -> Iterable:
     def prepare_data(self) -> None:
         # on multi-gpu jobs we only want to manipulate (download, etc) on node_rank=0, local_rank=0
         # or in the case where each node needs to do its own manipulation in which case just local_rank=0
-        if self.can_prepare_data():
-            if self.trainer.datamodule is not None:
+        local_rank_zero = self.trainer.local_rank == 0
+        global_rank_zero = self.trainer.local_rank == 0 and self.trainer.node_rank == 0
+
+        datamodule = self.trainer.datamodule
+        lightning_module = self.trainer.lightning_module
+        # handle datamodule prepare data:
+        # check for prepare_data_per_node & datamodule lifecycle properties before calling datamodule.prepare_data
+        if datamodule is not None and not datamodule.has_prepared_data:
+            dm_prepare_data_per_node = datamodule.prepare_data_per_node
+            if (self.trainer.prepare_data_per_node is not None) and (
+                datamodule.prepare_data_per_node != self.trainer.prepare_data_per_node
+            ):
+                raise MisconfigurationException(
+                    f"Inconsistent settings found for `prepare_data_per_node`. "
+                    f"Value was set with both `Trainer(prepare_data_per_node={self.trainer.prepare_data_per_node}.)` "
+                    f"and `DataModule.prepare_data_per_node={datamodule.prepare_data_per_node}`. "
+                    f"Move `prepare_data_per_node` setting to DataModule property"
+                )
+            if (dm_prepare_data_per_node and local_rank_zero) or (not dm_prepare_data_per_node and global_rank_zero):
                 self.trainer.datamodule.prepare_data()
-            self.trainer.lightning_module.prepare_data()
-            self.trainer._is_data_prepared = True
-
-    def can_prepare_data(self):
-        should_call_dm_prepare_data = True
-        if self.trainer.datamodule is not None and is_overridden("prepare_data", self.trainer.datamodule):
-            should_call_dm_prepare_data = not self.trainer.datamodule._has_prepared_data
-
-        if self.trainer.prepare_data_per_node:
-            return self.trainer.local_rank == 0 and should_call_dm_prepare_data
-        return self.trainer.node_rank == 0 and self.trainer.local_rank == 0 and should_call_dm_prepare_data
+        # handle lightning module prepare data:
+        # check for prepare_data_per_node before calling lightning_module.prepare_data
+        if lightning_module is not None:
+            lm_prepare_data_per_node = lightning_module.prepare_data_per_node
```

Comment:
@SeanNaren @tchaton - one test is failing for deepspeed due to this line
https://dev.azure.com/PytorchLightning/pytorch-lightning/_build/results?buildId=25289&view=logs&j=3afc50db-e620-5b81-6016-870a6976ad29&t=d9f671c5-a304-5675-5394-961fd7f98b9b

n00b question: for deepspeed, why isn't `trainer.lightning_module` not actually a lightning module here? it's showing that `lightning_module` is `ModelParallelClassificationModel` ?